cpp/include/cudf/contiguous_split.hpp

```diff
@@ -127,6 +127,148 @@ std::vector<packed_table> contiguous_split(
+ * //
+ * // Internally, the GPU unit-of-work is a 1MB batch. When we instantiate `cudf::chunked_pack`,
+ * // all the 1MB batches for the source table_view are computed up front. Additionally,
+ * // chunked_pack calculates the number of iterations that are required to go through all those
+ * // batches given a `user_buffer_size` buffer. The number of 1MB batches in each iteration (chunk)
+ * // equals the number of CUDA blocks that will be used for the main kernel launch.
+ * //
+ * std::size_t user_buffer_size = 128*1024*1024;
+ *
+ * auto chunked_packer = make_chunked_pack(tv, user_buffer_size, stream, mr);
+ *
+ * std::size_t host_offset = 0;
+ * auto host_buffer = ...; // obtain a host buffer you would like to copy to
+ *
+ * while (chunked_packer->has_next()) {
+ *   // get a user buffer of size `user_buffer_size`
+ *   cudf::device_span<uint8_t> user_buffer = ...;
+ *   std::size_t bytes_copied = chunked_packer->next(user_buffer);
+ *
+ *   // buffer will hold the contents of at most `user_buffer_size` bytes
+ *   // of the contiguously packed input `table_view`. You are now free to copy
+ *   // this memory somewhere else, for example, to host.
+ *   cudaMemcpyAsync(
+ *     host_buffer.data() + host_offset,
+ *     user_buffer.data(),
+ *     bytes_copied,
+ *     cudaMemcpyDefault,
+ *     stream);
+ *
+ *   host_offset += bytes_copied;
+ * }
+ */
+class chunked_pack {
+ public:
+  /**
+   * @brief Construct a `chunked_pack` class.
+   *
+   * @param input source `table_view` to pack
+   * @param user_buffer_size buffer size (in bytes) that will be passed on `next`. Must be
+   *                         at least 1MB
+   * @param stream CUDA stream used for device memory operations and kernel launches
+   * @param mr RMM memory resource to be used for temporary and scratch allocations only
+   */
+  explicit chunked_pack(cudf::table_view const& input,
+                        std::size_t user_buffer_size,
+                        rmm::cuda_stream_view stream,
+                        rmm::mr::device_memory_resource* mr);
+
+  /**
+   * @brief Destructor that will be implemented as default, required because
+   * contiguous_split_state is incomplete at this stage.
+   */
+  ~chunked_pack();
+
+  /**
+   * @brief Obtain the total size of the contiguously packed `table_view`.
+   *
+   * @return total size (in bytes) of all the chunks
+   */
+  [[nodiscard]] std::size_t get_total_contiguous_size() const;
+
+  /**
+   * @brief Function to check if there are chunks left to be copied.
+   *
+   * @return true if there are chunks left to be copied, and false otherwise
+   */
+  [[nodiscard]] bool has_next() const;
+
+  /**
+   * @brief Packs the next chunk into `user_buffer`. This should be call as long as
+   * `has_next` returns true. If `next` is called when `has_next` is false, an exception
+   * is thrown.
+   *
+   * @throws cudf::logic_error If the size of `user_buffer` is different than `user_buffer_size`
```

Comment:
This may be overkilled. In theory, we can unpack a chunk of preset data into any buffer that is large enough to hold the data.

Reply:
Imagine that we can have a re-usable buffer and we can unpack many tables with different `user_buffer_size` values.

Reply:
Yeah it is overkill, the real requirement is that the span passed here is bigger than what the next iteration needs. If there is a single iteration, and there isn't much to pack, locking up a large device buffer may be very wasteful.

I think I could add a `get_next_size()` method, and the requirement is that the buffer passed on next must be >= than `get_next_size()`. We can then call `user_buffer_size` something like `maximum_user_size`, and the user should be prepared to produce a buffer up to that size.

Reply:
I'd stick with keeping it simple for now.  This seems like a potential followup optimization if we find out it actually is an issue.  This PR is already has plenty of delta :)

Reply:
lets file this as a follow on, discussing with @nvdbaranec. We can make this as an improvement.

Reply:
https://github.com/rapidsai/cudf/issues/13332